Can't compile on React components on production with Vite but works on development

### Command

build

### Is this a regression?

- [x] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

webpack builder

### Description

I can use react components okay with vite with ng serve, but it won't build

My org is a multi front-end team, and some teams work with react, others with angular, and is a big system, we can't migrate all react to angular.

During serve everything works flawlessly but during build it fails on compile time



### Minimal Reproduction

[Create a new angular project with ssr

Add react, react dom ](https://github.com/alexandresebrao/ssr-19-angular-react)

This is a minimal repro.

### Exception or Error

✘ [ERROR] Expected ")" but found "Component"

    src/app/react.tsx:6:16:
      6 │       <h1>React Component</h1>
        │                 ~~~~~~~~~
        ╵                 )


### Your Environment

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 19.0.0-next.9
Node: 20.17.0
Package Manager: npm 10.9.0
OS: linux x64

Angular: 19.0.0-next.8
... animations, common, core, forms, platform-browser
... platform-browser-dynamic, platform-server, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1900.0-next.9
@angular-devkit/build-angular   19.0.0-next.9
@angular-devkit/core            19.0.0-next.9
@angular-devkit/schematics      19.0.0-next.9
@angular/cli                    19.0.0-next.9
@angular/compiler               19.0.0-next.9
@angular/compiler-cli           19.0.0-next.9
@angular/ssr                    19.0.0-next.9
@schematics/angular             19.0.0-next.9
rxjs                            7.8.1
typescript                      5.6.2
zone.js                         0.15.0


### Anything else relevant?

I think if it doesn't have a problem during serve, it should work okay with build.

--- a/packages/angular/build/src/builders/application/tests/behavior/typescript-isolated-modules_spec.ts
+++ b/packages/angular/build/src/builders/application/tests/behavior/typescript-isolated-modules_spec.ts
@@ -47,5 +47,33 @@ describeBuilder(buildApplication, APPLICATION_BUILDER_INFO, (harness) => {
 
       expect(result?.success).toBe(true);
     });
+
+    it('supports TSX files with isolated modules enabled and enabled optimizations', async () => {
+      // Enable tsconfig isolatedModules option in tsconfig
+      await harness.modifyFile('tsconfig.json', (content) => {
+        const tsconfig = JSON.parse(content);
+        tsconfig.compilerOptions.isolatedModules = true;
+        tsconfig.compilerOptions.jsx = 'react-jsx';
+
+        return JSON.stringify(tsconfig);
+      });
+
+      await harness.writeFile('src/types.d.ts', `declare module 'react/jsx-runtime' { jsx: any }`);
+      await harness.writeFile('src/abc.tsx', `export function hello() { return <h1>Hello</h1>; }`);
+      await harness.modifyFile(
+        'src/main.ts',
+        (content) => content + `import { hello } from './abc'; console.log(hello());`,
+      );
+
+      harness.useTarget('build', {
+        ...BASE_OPTIONS,
+        optimization: true,
+        externalDependencies: ['react'],
+      });
+
+      const { result } = await harness.executeOnce();
+
+      expect(result?.success).toBe(true);
+    });
   });
 });

--- a/packages/angular/build/src/tools/esbuild/angular/compiler-plugin.ts
+++ b/packages/angular/build/src/tools/esbuild/angular/compiler-plugin.ts
@@ -8,6 +8,7 @@
 
 import type {
   BuildFailure,
+  Loader,
   Metafile,
   OnStartResult,
   OutputFile,
@@ -456,9 +457,21 @@ export function createCompilerPlugin(
           typeScriptFileCache.set(request, contents);
         }
 
+        let loader: Loader;
+        if (useTypeScriptTranspilation || isJS) {
+          // TypeScript has transpiled to JS or is already JS
+          loader = 'js';
+        } else if (request.at(-1) === 'x') {
+          // TSX and TS have different syntax rules. Only set if input is a TSX file.
+          loader = 'tsx';
+        } else {
+          // Otherwise, directly bundle TS
+          loader = 'ts';
+        }
+
         return {
           contents,
-          loader: useTypeScriptTranspilation || isJS ? 'js' : 'ts',
+          loader,
         };
       });
 

